Sources/Equ/ElementwiseSequenceEqualityComparer.cs

```diff
@@ -13,12 +14,63 @@
     public class ElementwiseSequenceEqualityComparer<T> : EqualityComparer<T> where T : IEnumerable
     {
+        private static readonly Type EnumerableType = typeof(T)
+            .GetTypeInfo()
+            .GetInterfaces()
+            .Where(type => type.GetTypeInfo().IsGenericType && typeof(IEnumerable<>).GetTypeInfo().IsAssignableFrom(type.GetGenericTypeDefinition()))
+            .SelectMany(type => type.GetTypeInfo().GetGenericArguments())
+            .SingleOrDefault();
+
+        private static readonly MethodInfo SequenceEqualsMethodInfo = (EnumerableType == null) ? null : typeof(Enumerable)
+            .GetTypeInfo()
+            .GetMethods(BindingFlags.Static | BindingFlags.Public)
+            .Where(methodInfo => methodInfo.Name.Equals(nameof(Enumerable.SequenceEqual)) && methodInfo.GetParameters().Length == 3)
+            .Single()
+            .MakeGenericMethod(EnumerableType);
+
+        private static readonly MethodInfo ScrambledEqualsMethodInfo = (EnumerableType == null) ? null : typeof(ElementwiseSequenceEqualityComparer<T>)
+            .GetTypeInfo()
+            .GetMethods(BindingFlags.Static | BindingFlags.NonPublic)
+            .Where(methodInfo => methodInfo.Name.Equals(nameof(ElementwiseSequenceEqualityComparer<T>.ScrambledEquals)) && methodInfo.IsGenericMethod)
+            .Single()
+            .MakeGenericMethod(EnumerableType);
+
         // ReSharper disable once UnusedMember.Global
-        public new static ElementwiseSequenceEqualityComparer<T> Default => new ElementwiseSequenceEqualityComparer<T>();
+        public new static ElementwiseSequenceEqualityComparer<T> Default => new ElementwiseSequenceEqualityComparer<T>(MemberwiseEqualityMode.None);
+        public static ElementwiseSequenceEqualityComparer<T> ByFieldsRecursive => new ElementwiseSequenceEqualityComparer<T>(MemberwiseEqualityMode.ByFieldsRecursive);
+        public static ElementwiseSequenceEqualityComparer<T> ByPropertiesRecursive => new ElementwiseSequenceEqualityComparer<T>(MemberwiseEqualityMode.ByPropertiesRecursive);
 
         // ReSharper disable once StaticMemberInGenericType
         private static readonly bool _typeHasDefinedOrder = !IsDictionaryType() && !IsSetType();
-        
+
+        private bool _recursive;
+        private Lazy<object> _memberwiseEqualityComparer;
+
+        public ElementwiseSequenceEqualityComparer() : this(MemberwiseEqualityMode.None) { }
+
+        internal ElementwiseSequenceEqualityComparer(MemberwiseEqualityMode mode)
+        {
+            _recursive = (mode == MemberwiseEqualityMode.ByFieldsRecursive || mode == MemberwiseEqualityMode.ByPropertiesRecursive);
+            _memberwiseEqualityComparer = new Lazy<object>(() => CreateMemberwiseEqualityComparer(mode));
+        }
+
+        private object CreateMemberwiseEqualityComparer(MemberwiseEqualityMode mode)
+        {
+            var propertyName = mode switch

```

Comment:
I don't think I've seen this syntax before. I haven't paid much attention to the new features for a few years though. What is it doing?

Reply:
It's C# 8.0's pattern matching: https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/may/csharp-8-0-pattern-matching-in-csharp-8-0